### PR TITLE
Tenant & API key creation

### PR DESCRIPTION
**Pull request by memosys**
Added basic login process based on generated secret key;

todo: for some reason NEST Server closes if endpoints are called from browser maybe a CORS issue;

this PR tackles following issues

1. #12 
2. #14 (partial- create tenant/secret key and login to dashboard)
todo: API Key creation
3. Makes some necessary updates to the db schema as discussed offline

### CODE REVIEW

File: web-portal/backend/src/tenant/tenant.service.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  async countAll(): Promise<number> {
+    const count = await this.prisma.client.tenant.count();
+    return count;
+  }
+
+  async create(): Promise<any> {
+    const secretKey = randomBytes(8).toString('hex');
+    const salt = await bcrypt.genSalt();
+
+    const hashedKey = await bcrypt.hash(secretKey, salt);
```
Comment (plor) [inline web-portal/backend/src/tenant/tenant.service.ts:23]: I don't think this works for our usecase. We want the tenant secret to be both the lookup key and hashed to secure it. This method (as seen below) requires a separate lookup key so the comparison only has to be done once. Is there a way to either use a shared salt so the hash is deterministic?

**Review by plor — CHANGES_REQUESTED** [review]
Most of my comments are just things to consider moving forward, but I'd like to discuss REST a bit to make sure we're on the same page.

**Comment by memosys** [comment]
@plor I requested review earlier today after working out feedback, feel free to review but not intending a merge yet;
Intending to cover #12 #13 #14 completely in this PR now! let me know if you think otherwise!

File: web-portal/backend/src/tenant/tenant.controller.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+@Controller('tenant')
+export class TenantController {
+  constructor(private readonly tenantService: TenantService) {}
+
+  @Get('count')
+  async generateTenant(): Promise<any> {
+    console.log('This action returns a count of total tenants');
+    return this.tenantService.countAll();
+  }
```
Comment (plor) [inline web-portal/backend/src/tenant/tenant.controller.ts:13]: I'm just wondering why count is needed?

File: web-portal/backend/src/authkeys/authkeys.controller.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { Controller, Put, Query, Param } from '@nestjs/common';
+import { AuthkeysService } from './authkeys.service';
+
+@Controller('authkeys')
+export class AuthkeysController {
+  constructor(private readonly authKeyService: AuthkeysService) {}
+
+  @Put(':tenantId/create')
```
Comment (plor) [inline web-portal/backend/src/authkeys/authkeys.controller.ts:8]: This should be a POST and remove the create from the path. I would suggest `/` for the path with the tenantId in the post data. The other option would be to figure out some sort of subcontroller, so the path could be `/tenant/:tenantId/authkey/`

**Review by plor — CHANGES_REQUESTED** [review]
Made a couple comments, but it is hard to review when additional comments are added to the PR. I recommend smaller PRs that we merge more often. Having 1 PR for multiple tasks either means the tasks are created incorrectly, or the dev flow isn't working for us.

File: web-portal/backend/src/tenant/tenant.controller.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    return validation;
+  }
+
+  @Get('/:id')
+  async getTenantById(@Param('id') id: string): Promise<any> {
+    console.log('This action returns a tenant by its id');
+    return this.tenantService.getTenantById(id);
+  }
+
+  @Get('applyCredits/:id')
```
[the remark this answers is not in the document]
Comment (plor) [reply web-portal/backend/src/tenant/tenant.controller.ts]: I realized that this should be POST as it isn't idempotent.

File: web-portal/backend/src/tenant/tenant.controller.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    return this.tenantService.countAll();
+  }
+
+  @Get('create')
+  async createTenant(): Promise<any> {
+    console.log('This action creates a new tenant');
+    return this.tenantService.create();
+  }
+
+  @Get('validate/:key')
```
[the remark this answers is not in the document]
Comment (plor) [reply web-portal/backend/src/tenant/tenant.controller.ts]: Another question, do we need a validate, or is a normal get that doesn't return 404 the same?

File: web-portal/backend/src/tenant/tenant.controller.ts (hunk 2 above)
Comment (memosys) [reply web-portal/backend/src/tenant/tenant.controller.ts:13]: removing this, it was temporary for dev purpose only!

File: web-portal/backend/src/authkeys/authkeys.controller.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import { Controller, Post, Param } from '@nestjs/common';
+import { AuthkeysService } from './authkeys.service';
+
+@Controller('tenant')
+export class AuthkeysController {
+  constructor(private readonly authKeyService: AuthkeysService) {}
+
+  @Post(':tenantId/authkey')
+  async applyCredits(@Param('tenantId') tenantId: string): Promise<any> {
```
Comment (plor) [inline web-portal/backend/src/authkeys/authkeys.controller.ts:9]: This function name is confusing, should be `createAuthKey()`?

Comment (plor) [reply web-portal/backend/src/authkeys/authkeys.controller.ts:9]: Also, this should require the secretkey as a parameter to authorize this action. The confusion between the tenantId and the secret key should be rectified as I was stuck on the secret being the primary key, but either way it needs to be clear where the authentication/authorization lies.

**Review by plor — APPROVED** [review]
Just a couple points of confusion, and we need to make sure to protect the endpoints properly.

**Pull request merged**